sdk/core/System.ClientModel/src/Convenience/AsyncPageCollectionOfT.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ClientModel;
+
+/// <summary>
+/// An asynchronous collection of page results returned by a cloud service.
+/// Cloud services use pagination to return a collection of items over multiple
+/// responses.  Each response from the service returns a page of items in the
+/// collection, as well as the information needed to obtain the next page of
+/// items, until all the items in the requested collection have been returned.
+/// To enumerate the items in the collection, instead of the pages in the
+/// collection, call <see cref="GetAllValuesAsync"/>.  To get the current
+/// collection page, call <see cref="GetCurrentPageAsync"/>.
+/// </summary>
+public abstract class AsyncPageCollection<T> : IAsyncEnumerable<PageResult<T>>
+{
+    /// <summary>
+    /// Create a new instance of <see cref="AsyncPageCollection{T}"/>.
+    /// </summary>
+    protected AsyncPageCollection() : base()
+    {
+        // Note that page collections delay making a first request until either
+        // GetCurrentPageAsync is called or the collection returned by
+        // GetAllValuesAsync is enumerated, so this constructor calls the base
+        // class constructor that does not take a PipelineResponse.
+    }
+
+    /// <summary>
+    /// Get the current page of the collection.
+    /// </summary>
+    /// <returns>The current page in the collection.</returns>
+    public async Task<PageResult<T>> GetCurrentPageAsync()
+        => await GetCurrentPageAsyncCore().ConfigureAwait(false);
+
+    /// <summary>
+    /// Get a collection of all the values in the collection requested from the
+    /// cloud service, rather than the pages of values.
+    /// </summary>
+    /// <returns>The values requested from the cloud service.</returns>
+    public async IAsyncEnumerable<T> GetAllValuesAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        await foreach (PageResult<T> page in this.WithCancellation(cancellationToken).ConfigureAwait(false))
+        {
+            foreach (T value in page.Values)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
```

Comment:
I wish there was a way to avoid doing this check on each iteration, but I can't think of one without weakening the contract.

Reply:
We could check it before every service call instead -- I had thought it might be better to cancel more quickly if someone cancelled the token though.  Do you feel strongly?  Is the cost of the check high?

Reply:
It's a `volatile` read, so there's some overhead to each call but not enough that I'd say it outweighs honoring the token more quickly.   I agree with your thought - unless we see something perf-critical, we should leave as-written, unless someone has a clever idea.    

Pragmatically, the caller is enumerating pages/results to perform some kind of processing which quite likely dominates the time spent per-item.